service-worker/index.js

```diff
@@ -37,24 +39,68 @@ self.addEventListener('fetch', (event) => {
   let scopeExcluded = urlMatchesAnyPattern(request.url, INDEX_EXCLUDE_SCOPE);
   let scopeIncluded = !INDEX_INCLUDE_SCOPE.length || urlMatchesAnyPattern(request.url, INDEX_INCLUDE_SCOPE);
   let isTests = url.pathname === '/tests' && ENVIRONMENT === 'development';
-
   if (!isTests && isGETRequest && isHTMLRequest && isLocal && scopeIncluded && !scopeExcluded) {
-    event.respondWith(
-      caches.match(INDEX_HTML_URL, { cacheName: CACHE_NAME })
-        .then((response) => {
-          if (response) {
-            return response;
-          }
-
-          // Re-fetch the resource in the event that the cache had been cleared
-          // (mostly an issue with Safari 11.1 where clearing the cache causes
-          // the browser to throw a non-descriptive blank error page).
-          return fetch(INDEX_HTML_URL, { credentials: 'include' })
-            .then((fetchedResponse) => {
-              caches.open(CACHE_NAME).then((cache) => cache.put(INDEX_HTML_URL, fetchedResponse));
-              return fetchedResponse.clone();
-            });
-        })
-    );
+    if (STRATEGY === 'fallback') {
+      cacheFallbackFetch(event, TIMEOUT);
+    }
+    else {
+      return cacheFirstFetch(event);
+    }
   }
 });
+
+function cacheFirstFetch(event) {
+  return event.respondWith(
+    caches.match(INDEX_HTML_URL, { cacheName: CACHE_NAME })
+      .then((response) => {
+        if (response) {
+          return response;
+        }
+
+        /**
+          Re-fetch the resource in the event that the cache had been cleared
+          (mostly an issue with Safari 11.1 where clearing the cache causes
+          the browser to throw a non-descriptive blank error page).
+        */
+        return fetch(INDEX_HTML_URL, { credentials: 'include' })
+          .then((fetchedResponse) => {
+            caches.open(CACHE_NAME).then((cache) => cache.put(INDEX_HTML_URL, fetchedResponse));
+            return fetchedResponse.clone();
+          });
+      })
+  );
+}
+
+function cacheFallbackFetch(event, fetchTimeout) {
+  const FETCH_TIMEOUT = fetchTimeout;
+  let didTimeOut = false;
+  new Promise(function(resolve, reject) {
```

Comment:
Does this also need a `return` here?

Reply:
Given it is working I suppose you do not. Seems like is some future cleanup here to make the two approaches behave more similarly - but not within the scope of this PR.